### PR TITLE
feat(eslint-plugin): [no-unnecessary-parameter-property-assignment] no unnecessary property assignment

### LINKED ISSUE #7045: Rule proposal: forbid unnecessary assignment of parameter properties
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

My rule would check that there are no unnecessary assignments of parameter properties in constructors.

### Fail Cases

```typescript
class Foo {
  constructor(public bar: string) {
    this.bar = bar; // <-- warn here
  }
}

class Foo2 {
  baz: string;
  constructor(public bar: string, baz: string) {
    this.bar = bar; // <-- warn here
    this.baz = baz;
  }
}
```

### Pass Cases

```typescript
class Foo {
  constructor(public bar: string) {}
}

class Foo2 {
  baz: string;
  constructor(public bar: string, baz: string) {
    this.baz = baz;
  }
}
```

### Additional Info

You can see in [this TypeScript playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiD29oG8BQ1rHgOwgFwCcBXYPeAgCgAciAjEAS2GlrAIC5p8CHsBzAJQp0GaHgAWDCADpWBaAF4WbANwiAvqk1A) that assignment of parameter properties produces duplicate assignment:

image

### PR DESCRIPTION
**Pull request by undsoft**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7045
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Adds new linter rule to check for useless assignment of constructor property to itself.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    const parameterProperties = new Set<string>();
+    return {
+      'ClassDeclaration, ClassExpression'(): void {
+        parameterProperties.clear();
+      },
+      'MethodDefinition[kind="constructor"] > FunctionExpression > TSParameterProperty'(
+        property: TSESTree.TSParameterProperty,
+      ): void {
+        if (property.parameter.type !== AST_NODE_TYPES.Identifier) {
+          return;
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts:35]: [Bug] Hmm, I think this is a case we'll want to tackle. Why limit to just identifiers?

```ts
class A {
  constructor(b: 'c' | 'd') {
    this.b = b;
  }
}
```

Suggestion: remove this `if` and the `Identifier` check later?

File: packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        }
+
+        parameterProperties.add(property.parameter.name);
+      },
+      'MethodDefinition[kind="constructor"] > FunctionExpression AssignmentExpression'(
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts:40]: [Cleanup] I think you can simplify some of the logic in this function:

```suggestion
      'MethodDefinition[kind="constructor"] > FunctionExpression > AssignmentExpression'(
```

This edge case comes to mind:

```ts
class A {
  constructor() {
    const b = function () {
      this.c = 'd';
    }
  }
}
```

_(I haven't verified this myself - good luck!)_

File: packages/eslint-plugin/tests/rules/no-unnecessary-parameter-property-assignment.test.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/no-unnecessary-parameter-property-assignment.test.ts:1]: [Tests] The test cases in this file do a great job of testing the case of a single class and various name situations (nice!). But there are a _ton_ of -sometimes horrifying- edge cases. Consider that this -definitely horrifying- code is valid TypeScript the rule has to handle:

```ts
class Base {
  constructor(public a: number) { }
}

class Derived extends class extends Base { constructor(public b: number) { super(b); this.b = b; } } {
  constructor(public c: number) {
    super(c);
    this.c = c;
  }
}
```

Suggestion: give the `.ts` test case files under `cases/conformance/classes` in https://github.com/microsoft/TypeScript/pull/29374/files a skim. This PR doesn't need to handle nuances around `super()` ordering - that PR just has a lot of nifty examples of odd edge cases for scoping along with how TypeScript's source code handles them internally.

File: packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import * as util from '../util';
+
+export const messageId = 'noUnnecessaryParameterPropertyAssignment';
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts:6]: [Style] We don't generally do this. We normally go for small clean message IDs. The practice of storing them in variables doesn't scale well to rules that have many messages.

```suggestion
```

File: packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    type: 'suggestion',
+    docs: {
+      description:
+        'Disallow unnecessary assignment of constructor property parameter to itself',
+      recommended: 'error',
+      requiresTypeChecking: false,
+    },
+    schema: [],
+    messages: {
+      [messageId]:
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts:20]: Maybe ... `unnecessaryAssignment`?

```suggestion
      unnecessaryAssignment:
```

File: packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        'Assignment has no effect as roperty parameter is already defined in the constructor.',
+    },
+  },
+  defaultOptions: [],
+  create(context) {
+    const parameterProperties = new Set<string>();
+    return {
+      'ClassDeclaration, ClassExpression'(): void {
+        parameterProperties.clear();
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts:29]: [Bug] Clearing `parameterProperties` on each new class won't hold up for inner classes declared inside outer classes. Consider this unfortunate bit of code:

```ts
class First {
  constructor(public a: number) {
    class Second {
      constructor(public a: number) {
        this.a = a;
      }
    }

    new Second(0);

    this.a = a;
  }
}
```

...and also imagine if there are variations of that where either or both `a`s are renamed to `b` and/or not followed with an unnecessary parameter property assignment.

Perhaps keeping a stack of the current class would be helpful, similar to how `no-invalid-this` has a `thisIsValidStack: boolean[]`?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Thanks for sending this in @undsoft! A solid start - but will need to handle some potentially gnarly edge cases. LMK if the feedback doesn't make sense!

The pirate announcer guy from Spongebob with his parrot, saying 'This is going to be great.'

**Comment by JoshuaKGoldberg** [comment]
👋 Hey @undsoft! Just checking in, is this still something you have time for? No worries if not - I just don't want to leave it hanging.

**Comment by JoshuaKGoldberg** [comment]
No worries, thanks for letting me know!

If anybody wants to drive it forward, please do post your own PR - and if you use this as a start, consider adding `Co-authored-by: @undsoft` at the end of your PR description. Thanks! 😊
